oca/api.py

```diff
@@ -49,15 +32,66 @@ def method(self):
 
             [1, 2, 3]
 
+        If no container is provided as an argument to the decorator, and a
+        value is returned from the decorated method, a ValueError will be
+        raised.
+
+        .. code-block:: python
+
+            @api.foreach
+            def method(self):
+                # Raises ValueError
+                return 1+2
+
+            @api.foreach
+            def method(self):
+                # No error raised
+                self.math = 1+2
+
+    Args:
+        container (type): Type of container to wrap results in. It will be
+            instantiated with an iterable of the decorated method results.
+
+    Raises:
+        ValueError: If no container type is provided, and a value is returned
+            from the decorated method.
+        TypeError: If an incorrect data type is provided as the `container`.
+            Valid arguments are `None`, instances of `type`, or a callable.
+
     Returns:
         list: Of the return values from the decorated method, in the same
             order as the original ``self`` RecordSet.
     """
 
-    def loop(method, self, *args, **kwargs):
-        result = [method(rec, *args, **kwargs) for rec in self]
-        return aggregate(method, result, self)
-
-    wrapper = decorator(loop, method)
-    wrapper._api = 'foreach'
-    return wrapper
+    method = None
+
+    if container is not None and not isinstance(container, type):
+        if callable(container):
+            method = container
+            container = None
+        else:
+            raise TypeError(
+                'An incorrect argument was provided to the `foreach` api.',
+            )
+
+    def _foreach(method):
+
+        def loop(method, self, *args, **kwargs):
+            results = [method(rec, *args, **kwargs) for rec in self]
```

Comment:
I think it's better to use a generator, because the `container` could be a lazy one (i.e. [`iter`](https://docs.python.org/2.7/library/functions.html#iter)), to improve performance under some circumstances, and you would be killing it here.

So, summarizing, I'd try to be as less smart as possible:

```
generator = (method(rec, *args, **kwargs) for rec in self)
if container is None:
    for result in generator:
        assert result is None
else:
    return container(generator)
```